File: grid75/graphs/01Matrix.cpp
// https://leetcode.com/problems/01-matrix/

#include <bits/stdc++.h>
using namespace std;

class Solution {
 public:
  bool isValid(int i, int j, vector<vector<int>>& mat) {
    return (i >= 0 and j >= 0 and i < mat.size() and j < mat[0].size());
  }

  // bfs approach
  vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
    int rows = mat.size();
    if (rows == 0)
      return mat;
    int cols = mat[0].size();
    vector<vector<int>> dist(rows, vector<int>(cols, INT_MAX));
    queue<pair<int, int>> q;
    for (int i = 0; i < rows; i++) {
      for (int j = 0; j < cols; j++) {
        if (mat[i][j] == 0) {
          dist[i][j] = 0;
          q.push({i, j});  // Put all 0s in the queue.
        }
      }
    }

    int dir[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    while (!q.empty()) {
      pair<int, int> curr = q.front();
      q.pop();
      for (int i = 0; i < 4; i++) {
        int new_r = curr.first + dir[i][0], new_c = curr.second + dir[i][1];
        if (new_r >= 0 && new_c >= 0 && new_r < rows && new_c < cols) {
          if (dist[new_r][new_c] > dist[curr.first][curr.second] + 1) {
            dist[new_r][new_c] = dist[curr.first][curr.second] + 1;
            q.push({new_r, new_c});
          }
        }
      }
    }
    return dist;
  }

  // dp approach
  vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
    int rows = mat.size();
    if (rows == 0)
      return mat;
    int cols = mat[0].size();
    vector<vector<int>> dist(rows, vector<int>(cols, INT_MAX - 100000));

    // First pass: check for left and top
    for (int i = 0; i < rows; i++) {
      for (int j = 0; j < cols; j++) {
        if (mat[i][j] == 0) {
          dist[i][j] = 0;
        } else {
          if (i > 0)
            dist[i][j] = min(dist[i][j], dist[i - 1][j] + 1);
          if (j > 0)
            dist[i][j] = min(dist[i][j], dist[i][j - 1] + 1);
        }
      }
    }

    // Second pass: check for bottom and right
    for (int i = rows - 1; i >= 0; i--) {
      for (int j = cols - 1; j >= 0; j--) {
        if (i < rows - 1)
          dist[i][j] = min(dist[i][j], dist[i + 1][j] + 1);
        if (j < cols - 1)
          dist[i][j] = min(dist[i][j], dist[i][j + 1] + 1);
      }
    }
    return dist;
  }
};

File: grind75_python/01Matrix.py
# https://leetcode.com/problems/01-matrix/

from typing import List
from queue import Queue
import sys


class Solution:
  def valid(self, i, j, mat):
    if 0 <= i < len(mat) and 0 <= j < len(mat[0]):
      return True
    return False

  def updateMatrixBFS(self, mat: List[List[int]]) -> List[List[int]]:

    ans = [[0 for _ in range(len(mat[0]))] for _ in range(len(mat))]

    directions = [(-1, 0), (0, -1), (1, 0), (0, 1)]

    q = Queue()
    visited = set()

    # add all the 0s to the q. we will do bfs from all the 0s and as soon as we encounter a 1 we will put the distance
    for i in range(len(mat)):
      for j in range(len(mat[0])):
        if mat[i][j] == 0:
          q.put((i, j))
          visited.add((i, j))

    dist = 0

    while not q.empty():

      # for each level away from a 0 we increment the distance
      for _ in range(q.qsize()):

        i, j = q.get()

        if mat[i][j] == 1:
          ans[i][j] = dist

        for d in directions:
          new_i = i+d[0]
          new_j = j+d[1]

          # we will add all the surrounding 1s. surrounding 0s will not be added because they are already in the visited set.
          if self.valid(new_i, new_j, mat) and (new_i, new_j) not in visited:
            q.put((new_i, new_j))
            visited.add((new_i, new_j))

      dist += 1

    return ans

  def updateMatrixDP(self, mat: List[List[int]]) -> List[List[int]]:

    ans = [[sys.maxsize for _ in range(len(mat[0]))] for _ in range(len(mat))]

    for i in range(len(mat)):
      for j in range(len(mat[0])):
        if mat[i][j] == 0:
          ans[i][j] = 0
        else:
          up = ans[i-1][j] if self.valid(i-1, j, mat) else sys.maxsize
          left = ans[i][j-1] if self.valid(i, j-1, mat) else sys.maxsize
          ans[i][j] = 1 + min(up, left)

    for i in range(len(mat) - 1, -1, -1):
      for j in range(len(mat[0]) - 1, -1, -1):
        down = ans[i+1][j] if self.valid(i+1, j, mat) else sys.maxsize
        right = ans[i][j+1] if self.valid(i, j+1, mat) else sys.maxsize
        ans[i][j] = min(ans[i][j], 1 + min(down, right))

    return ans


s = Solution()
# print(s.updateMatrix([[1, 1, 1], [1, 1, 1], [1, 1, 0]]))
